File: trunk/ShadowFrameworkJS/SF20Core/src/shadow/geometry/functions/SFCurvedTubeFunction.js

function SFCurvedTubeFunction(){
}

SFCurvedTubeFunction.prototype = {

	evalCurve:function(v, pts, P, dPdv){
	int v_index=(int)(v*pts.size());//Warning: Not well Identified 
	if(v_index==pts.size())			v_index--;//Warning: Not well Identified 
		 float  t = (v*pts.size())-v_index;
		 if ( v_index==0 ){
		 SFVertex3f  A = pts.get(0);
		 SFVertex3f  B = SFVertex3f.middle(pts.get(0),pts.get(1));
		P.set3f(A);
		P.mult(1-t);
		P.addMult3f(t, B);
		dPdv.set3f(B);
		dPdv.subtract3f(A);
	//d2Pdv2.set(0,0,0);//Warning: Not well Identified 
	}
	//else if(v_index==pts.size()-1);//Warning: Not well Identified 
		 SFVertex3f  A = SFVertex3f.middle(pts.get(v_index-1),pts.get(v_index));
		 SFVertex3f  B = pts.get(v_index);
		P.set3f(A);
		P.mult(1-t);
		P.addMult3f(t, B);
		dPdv.set3f(B);
		dPdv.subtract3f(A);
	//d2Pdv2.set(0,0,0);//Warning: Not well Identified 
	//}
		else{
		 SFVertex3f  A = SFVertex3f.middle(pts.get(v_index-1), pts.get(v_index));
		 SFVertex3f  B = pts.get(v_index);
		 SFVertex3f  C = SFVertex3f.middle(pts.get(v_index), pts.get(v_index+1));
		P.set3f(A);
		P.mult((1-t)*(1-t));
		P.addMult3f(2*t*(1-t), B);
		P.addMult3f(t*t, C);
		dPdv.set3f(A);
		dPdv.mult(-2*(1-t));
		dPdv.addMult3f(2-4*t, B);
		dPdv.addMult3f(2*t, C);
	//d2Pdv2.set(A);//Warning: Not well Identified 
		d2Pdv2.mult(2);
		d2Pdv2.addMult3f(-4,B);
		d2Pdv2.addMult3f(2, C);
	}
	},

	evalAll:function(v){
		 if ( v!=lastV ){
		evalCurve(v, Cc, Ccv, dCcdv);
		evalCurve(v, Rc, Rcv, new SFVertex3f());
		Vec1v.set3f(Rcv);
		Vec1v.subtract3f(Ccv);
		Vec2v.set3f(dCcdv.cross(Vec1v));
		Vec2v.normalize3f();
		Vec2v.mult((float)(Math.sqrt(Vec1v.dot3f(Vec1v))));
		lastV  = v;
	}
	},

	getX:function(u, v){
		evalAll(v);
		cos  = (float).(Math.cos(2*Math.PI*u));
		sin  = (float).(Math.sin(2*Math.PI*u));
		return ,Ccv.getX()+cos*Vec1v.getX()+sin*Vec2v.getX();
	},

	getY:function(u, v){
		return ,Ccv.getY()+cos*Vec1v.getY()+sin*Vec2v.getY();
	},

	getZ:function(u, v){
		return ,Ccv.getZ()+cos*Vec1v.getZ()+sin*Vec2v.getZ();
	},

	generateNewDatasetInstance:function(){
	// TODO Auto-generated method stub		return null;//Warning: Not well Identified 
	},

	getCode:function(){
	// TODO Auto-generated method stub		return null;//Warning: Not well Identified 
	},

	readFromStream:function(stream){
	// TODO Auto-generated method stub;//Warning: Not well Identified 
	},

	writeOnStream:function(stream){
	// TODO Auto-generated method stub;//Warning: Not well Identified 
	}

};

File: SFDataApplicationTestC++/shadow/geometry/functions/SFCurvedTubeFunction.cpp
/*
 * SFCurvedTubeFunction.cpp
 *
 *  Created on: 23/set/2013
 *      Author: Alessandro
 */

#include "shadow/geometry/functions/SFCurvedTubeFunction.h"

namespace sf {


SFCurvedTubeFunction::~SFCurvedTubeFunction(){
	delete centralCurve;
	delete rayCurve;
}


SFCurvedTubeFunction::SFCurvedTubeFunction() {
	lastV=-1;
	cos_=0;
	sin_=0;
	centralCurve=0;
	rayCurve=0;
}

SFCurvedTubeFunction::SFCurvedTubeFunction(SFCurve* centralCurve,
		SFCurve* rayCurve) {
	this->centralCurve = centralCurve;
	this->rayCurve = rayCurve;
	lastV=-1;
	cos_=0;
	sin_=0;
}

SFVertex3f SFCurvedTubeFunction::getDu() {
	SFVertex3f du(getDXDu(),getDYDu(),getDZDu());
	du.normalize3f();
	return du;
}

SFVertex3f SFCurvedTubeFunction::getDv() {
	SFVertex3f dv(getDXDv(),getDYDv(),getDZDv());
	dv.normalize3f();
	return dv;
}

SFVertex3f SFCurvedTubeFunction::getPosition() {
	return SFVertex3f(getX(),getY(),getZ());
}


void SFCurvedTubeFunction::evalAll(float v){
	if(v!=lastV){
		centralCurve->getVertex(v, &Ccv);
		centralCurve->getDevDt(v, &dCcdv);
		rayCurve->getVertex(v, &Vec1v);
		Vec1v.subtract3f(Ccv);
		SFVertex3f* cross=dCcdv.cross(Vec1v);
		Vec2v.setArray(cross->getV());
			delete cross;
		Vec2v.mult((float)(sqrt(Vec1v.dot3f(Vec1v)/Vec2v.dot3f(Vec2v))));

		SFVertex3f dCcdv2;
		centralCurve->getDev2Dt(v, &dCcdv2);
		rayCurve->getDevDt(v, &DVec1v);
		DVec1v.subtract3f(dCcdv);
		//DVec1v.normalize3f();
		cross=dCcdv.cross(DVec1v);
		DVec2v.setArray(cross->getV());
			delete cross;
			cross=dCcdv2.cross(DVec1v);
		DVec2v.add3f(dCcdv2);
			delete cross;
		lastV=v;
	}
}

float SFCurvedTubeFunction::getX(float u,float v) {
	evalAll(v);
	cos_=(float)(cos(2*M_PI*u));
	sin_=(float)(sin(2*M_PI*u));
	return Ccv.getX()+cos_*Vec1v.getX()+sin_*Vec2v.getX();
}

float SFCurvedTubeFunction::getY(float u,float v) {
	return Ccv.getY()+cos_*Vec1v.getY()+sin_*Vec2v.getY();
}

float SFCurvedTubeFunction::getZ(float u,float v) {
	return Ccv.getZ()+cos_*Vec1v.getZ()+sin_*Vec2v.getZ();
}


float SFCurvedTubeFunction::getX() {
	return Ccv.getX()+cos_*Vec1v.getX()+sin_*Vec2v.getX();
}

float SFCurvedTubeFunction::getY() {
	return Ccv.getY()+cos_*Vec1v.getY()+sin_*Vec2v.getY();
}

float SFCurvedTubeFunction::getZ() {
	return Ccv.getZ()+cos_*Vec1v.getZ()+sin_*Vec2v.getZ();
}

float SFCurvedTubeFunction::getDXDv() {
	return dCcdv.getX()+cos_*DVec1v.getX()+sin_*DVec2v.getX();
}

float SFCurvedTubeFunction::getDYDv() {
	return dCcdv.getY()+cos_*DVec1v.getY()+sin_*DVec2v.getY();
}

float SFCurvedTubeFunction::getDZDv() {
	return dCcdv.getZ()+cos_*DVec1v.getZ()+sin_*DVec2v.getZ();
}

float SFCurvedTubeFunction::getDXDu() {
	return -sin_*Vec1v.getX()+cos_*Vec2v.getX();
}

float SFCurvedTubeFunction::getDYDu() {
	return -sin_*Vec1v.getY()+cos_*Vec2v.getY();
}

float SFCurvedTubeFunction::getDZDu() {
	return -sin_*Vec1v.getZ()+cos_*Vec2v.getZ();
}

SFCurve* SFCurvedTubeFunction::getCentralCurve() {
	return centralCurve;
}

void SFCurvedTubeFunction::setCentralCurve(SFCurve* centralCurve) {
	this->centralCurve = centralCurve;
}

SFCurve* SFCurvedTubeFunction::getRayCurve() {
	return rayCurve;
}

void SFCurvedTubeFunction::setRayCurve(SFCurve* rayCurve) {
	this->rayCurve = rayCurve;
}

} /* namespace sf */
